### PR TITLE
outlier detector: verify outlier detector config after it has been initialized with defaults

### LINKED ISSUE #14535: Config rejection when outlier detection max_ejection_time is not set
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: *One line description*

*Description*: Config rejection "outlier detector's max_ejection_time cannot be smaller than base_ejection_time" when outlier detection max_ejection_time is not set https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/outlier_detection_impl.cc#L283

>What issue is being seen? Describe what should be happening instead of
the bug, for example: Envoy should not crash, the expected value isn't
returned, etc.
Envoy should not validate max_ejection_time if not set in config as max_ejection_time is a newly introduced configuration parameter.

*Repro steps*:
> Include sample requests, environment, etc. All data and inputs
required to reproduce the bug.
Config with base_ejection_time set but not max_ejection_time set fails validation.

outlier_detection {
      ...
      base_ejection_time {
        seconds: 30
      }
      ...
}

>**Note**: The [Envoy_collect tool](https://github.com/envoyproxy/envoy/blob/master/tools/envoy_collect/README.md)
gathers a tarball with debug logs, config and the following admin
endpoints: /stats, /clusters and /server_info. Please note if there are
privacy concerns, sanitize the data prior to sharing the tarball/pasting.

*Admin and Stats Output*:
>Include the admin output for the following endpoints: /stats,
/clusters, /routes, /server_info. For more information, refer to the
[admin endpoint documentation.](https://www.envoyproxy.io/docs/envoy/latest/operations/admin)

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Config*:
>Include the config used to configure Envoy.

*Logs*:
>Include the access logs and the Envoy logs.

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Call Stack*:
> If the Envoy binary is crashing, a call stack is **required**.
Please refer to the [Bazel Stack trace documentation](https://github.com/envoyproxy/envoy/tree/master/bazel#stack-trace-symbol-resolution).

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
Fixes bug where config values were validated before outlier detector was initialized with default values.

Additional Description:
Risk Level: Low
Testing: added unit test
Docs Changes: no
Release Notes: no
Platform Specific Features:
Fixes #14535

### CODE REVIEW

File: source/common/upstream/outlier_detection_impl.cc [hunk 1]
```diff
@@ -280,13 +280,14 @@ DetectorImpl::create(const Cluster& cluster,
                      const envoy::config::cluster::v3::OutlierDetection& config,
                      Event::Dispatcher& dispatcher, Runtime::Loader& runtime,
                      TimeSource& time_source, EventLoggerSharedPtr event_logger) {
-  if (config.max_ejection_time() < config.base_ejection_time()) {
+  std::shared_ptr<DetectorImpl> detector(
+      new DetectorImpl(cluster, config, dispatcher, runtime, time_source, event_logger));
+
+  if (detector->config().maxEjectionTimeMs() < detector->config().baseEjectionTimeMs()) {
```
Comment (hennna) [inline source/common/upstream/outlier_detection_impl.cc:286]: Would it also be possible to increase the default value of max ejection time? Currently there isn't a way to override the value using the v2 API.

Comment (cpakulski) [reply source/common/upstream/outlier_detection_impl.cc:286]: The main reason behind `max_ejection_time` was to limit how much this value may grow. It determines how quickly the node comes back to the pool.
V2 is fatal in 1.17 by default.
If you insist on using V2, you can overwrite max_ejection_time using runtime values: https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_runtime.html

Comment (mattklein123) [reply source/common/upstream/outlier_detection_impl.cc:286]: @hennna I'm surprised you want to increase the default. I think it's already pretty generous? In any case I agree that this is probably OK given that there is a runtime override for the field.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**